### PR TITLE
CBMC runtime analysis: Record pointer alias analysis metric

### PR DESCRIPTION
**Pull request by natasha-jeppu**
*Issue #, if available:*

*Description of changes:*
Generate pointer alias analysis report using points-to set metric information from CBMC. Takes json output obtained from cbmc --show-points-to-sets and generates a html report describing size and contents of the points-to sets for pointer dereference as recorded by CBMC.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cbmc_viewer/aliast.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+def get_alias_item(json_item):
+    pointer = json_item[JSON_POINTER_KEY]
+    json_item.pop(JSON_POINTER_KEY)
+    json_item.update({"numOfDeref": 1})
+
+    alias_item = {}
+    alias_item[pointer] = json_item
+
+    return alias_item
```
Comment (markrtuttle) [inline cbmc_viewer/aliast.py:142]: I would have written `return {pointer: json_item}`

File: cbmc_viewer/aliast.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+def is_equal(summary, item):
+    for key in item.keys():
+        if key in summary.keys():
+            is_equal_return = all([item[key][k] == summary[key][k] \
+                for k in item[key].keys() if k not in ['numOfDeref']])
```
Comment (markrtuttle) [inline cbmc_viewer/aliast.py:148]: Line continuation is not necessary, the (...) do the job.

File: cbmc_viewer/aliast.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    json_item.pop(JSON_POINTER_KEY)
+    json_item.update({"numOfDeref": 1})
+
+    alias_item = {}
+    alias_item[pointer] = json_item
+
+    return alias_item
+
+def is_equal(summary, item):
+    for key in item.keys():
```
Comment (markrtuttle) [inline cbmc_viewer/aliast.py:145]: What is this function doing?  It is not a predicate, because it has a side effect.  What is it checking for equality?

**Review by markrtuttle — APPROVED** [review]
I have some question I'd liked answered, but this is fine otherwise.

File: cbmc_viewer/aliast.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    },required=True)
+
+VALID_ALIAS = voluptuous.schema_builder.Schema({
+    str : VALID_ALIAS_ITEM # str here stores the pointer variable name
+    })
+
+VALID_ALIAS_SUMMARY = voluptuous.schema_builder.Schema({
+    'summary': VALID_ALIAS,
+    'max': int,
+    'total': int
```
[the remark this answers is not in the document]
Comment (natasha-jeppu) [reply cbmc_viewer/aliast.py:45]: Max is `max_points_to_set`.
Total is `total_points_to_set` calculated as sum of `points_to_set_size * number_of_deref`. This tracks pointer alias analysis complexity. For every data pointer dereference, CBMC refers to the points-to set to generate SSA expression.

File: cbmc_viewer/aliast.py (hunk 3 above)
Comment (natasha-jeppu) [reply cbmc_viewer/aliast.py:145]: The number of pointer dereferences is something that we calculate from the CBMC output. So dereferences are calculated as number of time CBMC refers to the points-to set for a given pointer. Every time CBMC refers to the points-to set, a corresponding `'pointer name, points-to set, points-to set size'` is displayed in CBMC json format. 
Here, we check if these fields match any existing parsed data, if yes, we increment number of dereferences, if not, we create a new data item in the dictionary.

File: cbmc_viewer/aliast.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+#   }
+# ]
+
+################################################################
+# Utility functions to generate alias summary
+
+def get_max_total(alias_summary):
+    max_val = 0
+    total = 0
+    for alias_item in alias_summary.items():
```
[the remark this answers is not in the document]
Comment (natasha-jeppu) [reply cbmc_viewer/aliast.py]: Fixed it.

File: cbmc_viewer/aliast.py (hunk 1 above)
Comment (natasha-jeppu) [reply cbmc_viewer/aliast.py]: Fixed.

File: cbmc_viewer/aliast.py (hunk 2 above)
Comment (natasha-jeppu) [reply cbmc_viewer/aliast.py:145]: Fixed.

File: cbmc_viewer/aliast.py (hunk 3 above)
Comment (markrtuttle) [reply cbmc_viewer/aliast.py:142]: A pointer may have different points-to sets in different locations.  We are counting the number of times a pointer with a given points-to set is dereferenced.

File: cbmc_viewer/aliast.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (markrtuttle) [reply cbmc_viewer/aliast.py]: The code size for any given pointer dereference increases from 1 to the size of the points-to set (the size of the case statement).

**Review by markrtuttle — COMMENTED** [review]
Notes from our discussion that Mark will fix.

**Pull request closed without merging**